rmw_fastrtps_shared_cpp/src/rmw_client.cpp

```diff
@@ -41,43 +38,82 @@ __rmw_destroy_client(
   rmw_node_t * node,
   rmw_client_t * client)
 {
-  rmw_ret_t ret = RMW_RET_OK;
+  rmw_ret_t final_ret = RMW_RET_OK;
   auto common_context = static_cast<rmw_dds_common::Context *>(node->context->impl->common);
   auto info = static_cast<CustomClientInfo *>(client->data);
   {
     // Update graph
     std::lock_guard<std::mutex> guard(common_context->node_update_mutex);
     rmw_gid_t gid = rmw_fastrtps_shared_cpp::create_rmw_gid(
-      identifier, info->request_publisher_->getGuid());
+      identifier, info->request_publisher_->guid());
     common_context->graph_cache.dissociate_writer(
       gid,
       common_context->gid,
       node->name,
       node->namespace_);
     gid = rmw_fastrtps_shared_cpp::create_rmw_gid(
-      identifier, info->response_subscriber_->getGuid());
+      identifier, info->response_subscriber_->guid());
     rmw_dds_common::msg::ParticipantEntitiesInfo msg =
       common_context->graph_cache.dissociate_reader(
       gid, common_context->gid, node->name, node->namespace_);
-    ret = rmw_fastrtps_shared_cpp::__rmw_publish(
+    final_ret = rmw_fastrtps_shared_cpp::__rmw_publish(
       identifier,
       common_context->pub,
       static_cast<void *>(&msg),
       nullptr);
   }
 
-  Domain::removeSubscriber(info->response_subscriber_);
-  Domain::removePublisher(info->request_publisher_);
-  delete info->pub_listener_;
-  delete info->listener_;
-  _unregister_type(info->participant_, info->request_type_support_);
-  _unregister_type(info->participant_, info->response_type_support_);
-  delete info;
+  /////
+  // Delete DataWriter and DataReader
+  auto participant_info =
+    static_cast<CustomParticipantInfo *>(node->context->impl->participant_info);
+
+  // NOTE: Topic deletion and unregister type is done in the participant
+  if (nullptr != info) {
+    // Delete DataReader
+    ReturnCode_t ret = participant_info->subscriber_->delete_datareader(info->response_subscriber_);
+    if (ret != ReturnCode_t::RETCODE_OK) {
+      RMW_SET_ERROR_MSG("Fail in delete datareader");
+      return rmw_fastrtps_shared_cpp::cast_error_dds_to_rmw(ret);
+    }
+
+    // Delete DataReader listener
+    if (nullptr != info->listener_) {
+      delete info->listener_;
+    }
+
+    // Delete DataWriter
+    ret = participant_info->publisher_->delete_datawriter(info->request_publisher_);
+    if (ret != ReturnCode_t::RETCODE_OK) {
+      RMW_SET_ERROR_MSG("Fail in delete datareader");
+      return rmw_fastrtps_shared_cpp::cast_error_dds_to_rmw(ret);
```

Comment:
```suggestion
    if (ret != ReturnCode_t::RETCODE_OK && final_ret != RMW_RET_OK) {
      RMW_SAFE_FWRITE_TO_STDERR(
        RCUTILS_STRINGIFY(__function__) ":" RCUTILS_STRINGIFY(__line__) ": "
        "couldn't publish reader disassociation from graph_cache when destroying client");
      RMW_SET_ERROR_MSG("Fail in delete datareader");
      return rmw_fastrtps_shared_cpp::cast_error_dds_to_rmw(ret);
```

See previous comment.